### DISCUSSION TITLE
Question:How memory allocated?

### DISCUSSION THREAD

**Discussion by UWFFA** [skypjack/entt#1152] [Q&A]
Let's say we have a Mesh class, which in turn has another class, Texture.
It looks like this:

`Class` `Mesh`
`{`  
`public:`  
 `//Millions of very interesting methods  `
`private:`  
`Texture * usage_texture; `
`};`  

I want to add a MeshComponent that looks something like this:

`struct` `MeshComponent`
`{`
`Mesh mesh;`
  `//constructor and etc`
`};`

Where should I declare variables in the class (should usage_texture be created via new/smart pointers or just convert to Texture usage_texture) if I want to create all heavy objects (Mesh, Texture etc) in Heap. MeshComponent will be added via entt::registry::emplace.

**Comment by skypjack** [comment]
How does this question relate to EnTT? Isn't this just an application design issue? 🤔

**Reply by UWFFA** [reply]
I just want to understand exactly how the memory will be allocated. As I understand it, the containers themselves (which we get via entt::registry::get) are arranged in a way that reduces memory jumps.  But what about the objects themselves that they store? Could there be a situation where I've got a MeshComponent, and I want to use a Mesh object (for example, to use another file as a texture in usage_texture), but I have to make a memory jumps to do so. I realise this example may look silly, but I really want to understand why using ECS leads to better performance. And so far I have come to 2 conclusions:
1) Containers allocate memory in some magical way.
2) The improvement pops up because we split our functionality into components (in the case of Mesh, we can create a bunch of other components like TransformComponent, TextureComponent, etc) and because of that they are not stored in the same class (let's say a Texture object takes 8 bytes and the class is 108. If we want to change the texture, we only need to access the object with 8 bytes, not 108).

**Reply by skypjack** [reply]
Roughyl speaking, it's easier to arrange iterations to take advantage of cache locality. This is enough to reduce the number of CPU cycles used for many operations. That said, ECS isn't some magical tool that gives you performance for free. There are many implementations out there that like to lie and convince you of this, but no, this won't happen. It's just a tool that can help if used correctly but it cannot solve the problems of bad design or address all the problems of software development. 🤷‍♂️
